core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -216,7 +219,9 @@ public void phase1(final IndexCommit snapshot, final Translog.View translogView)
                     }
                     totalSize += md.length();
                 }
-                for (StoreFileMetaData md : Iterables.concat(diff.different, diff.missing)) {
+                List<StoreFileMetaData> phase1Files = new ArrayList<>(diff.different);
```

Comment:
We can avoid arraylist resizing by creating it with `diff.different.size() + diff.missing.size()` and then doing `addAll` for the diff parts